File: engine/app/services/coplan/plans/merge_text.rb
require "diff/lcs"

module CoPlan
  module Plans
    # A three-way, character-range merge. Both sides are compared with the
    # same immutable base. Only disjoint edits (or identical edits) combine.
    class MergeText
      class Conflict < StandardError; end

      def self.hunks(before, after)
        result = []
        position = 0
        pending = nil
        Diff::LCS.sdiff(before.each_char.to_a, after.each_char.to_a).each do |change|
          if change.action == "="
            result << pending if pending
            pending = nil
            position += 1
          else
            pending ||= { from: position, to: position, text: "" }
            unless change.action == "+"
              position += 1
              pending[:to] = position
            end
            pending[:text] += change.new_element unless change.action == "-"
          end
        end
        result << pending if pending
        result
      end

      def self.overlap?(left, right)
        if left[:from] == left[:to] && right[:from] == right[:to]
          left[:from] == right[:from]
        elsif left[:from] == left[:to]
          left[:from] > right[:from] && left[:from] < right[:to]
        elsif right[:from] == right[:to]
          right[:from] > left[:from] && right[:from] < left[:to]
        else
          left[:from] < right[:to] && right[:from] < left[:to]
        end
      end

      def self.call(base:, local:, remote:)
        return remote if local == base || local == remote
        return local if remote == base
        ours = hunks(base, local)
        theirs = hunks(base, remote)
        ours.each do |left|
          theirs.each do |right|
            raise Conflict, "Both edits change the same passage" if left != right && overlap?(left, right)
          end
        end
        merged = base.dup
        (ours + theirs).uniq.sort_by { |hunk| [ hunk[:from], hunk[:to] ] }.reverse_each do |hunk|
          merged[hunk[:from]...hunk[:to]] = hunk[:text]
        end
        merged
      end
    end
  end
end

File: engine/app/javascript/coplan/merge_text.js
import { diffChars } from "diff"

export class MergeConflict extends Error {}
export function textHunks(before, after) {
  const result = []
  let position = 0, pending = null
  for (const part of diffChars(before, after)) {
    if (!part.added && !part.removed) {
      if (pending) result.push(pending)
      pending = null
      position += part.value.length
    } else {
      pending ||= { from: position, to: position, text: "" }
      if (part.removed) { position += part.value.length; pending.to = position }
      else pending.text += part.value
    }
  }
  if (pending) result.push(pending)
  return result
}
function overlaps(a, b) {
  if (a.from === a.to && b.from === b.to) return a.from === b.from
  if (a.from === a.to) return a.from > b.from && a.from < b.to
  if (b.from === b.to) return b.from > a.from && b.from < a.to
  return a.from < b.to && b.from < a.to
}
export function mergeText(base, local, remote) {
  if (local === base || local === remote) return remote
  if (remote === base) return local
  const ours = textHunks(base, local), theirs = textHunks(base, remote)
  const same = (a, b) => a.from === b.from && a.to === b.to && a.text === b.text
  for (const a of ours) for (const b of theirs) {
    if (!same(a, b) && overlaps(a, b)) throw new MergeConflict("Both edits change the same passage")
  }
  const edits = [...theirs, ...ours.filter(a => !theirs.some(b => same(a, b)))].sort((a, b) => b.from - a.from || b.to - a.to)
  return edits.reduce((text, edit) => text.slice(0, edit.from) + edit.text + text.slice(edit.to), base)
}
export function mergeField(base, local, remote, label) {
  if (local === base || local === remote) return remote
  if (remote === base) return local
  throw new MergeConflict(`Both edits change the ${label}`)
}
